calico_containers/pycalico/ipam.py

```diff
@@ -782,6 +820,56 @@ def get_assignment_attributes(self, address):
             _, attributes = block.get_attributes_for_ip(address)
             return attributes
 
+    def get_cidr_assignment_attributes(self, cidr):
+        """
+        Return the attributes of a given CIDR. Depending on the size of the
+        CIDR in question, this could have a very large result.
+
+        :param cidr: CIDR to query.
+        :type cidr: IPNetwork
+        :return: The attributes of all the addresses in a given CIDR.
+        :rtype: list
+        """
+        assert isinstance(cidr, IPNetwork)
+
+        # List of IPNetworks corresponding to blocks in ETCD
+        block_cidrs = self._get_all_blocks(cidr)
+
+        # List of AllocationBlocks
+        blocks = []
+        for block_cidr in block_cidrs:
+            block = self._read_block(block_cidr)
+            blocks.append(block)
+
+        # List of IPPools
+        pools = self.get_ip_pools(cidr.version)
+
+        result = []
+        for block in blocks:
+            for pool_candidate in pools:
+                if block in pool_candidate:
+                    pool = str(pool_candidate)
+                    break
+
+            for num, index in enumerate(block.allocations):
+                ip_int = block.first + num
+                if cidr.first <= ip_int <= cidr.last:
+                    ip_attrs = {}
+                    ip_attrs["POOL"] = pool
+                    if index is None:
+                        ip_attrs["ALLOCATED"] = False
+                        ip_attrs["ATTRIBUTES"] = {}
+                        ip_attrs["HANDLE"] = None
+                    else:
+                        attributes = block.attributes[index]
+                        ip_attrs["ATTRIBUTES"] = attributes["secondary"]
+                        ip_attrs["HANDLE"] = attributes["handle_id"]
+                        ip_attrs["IP"] = str(IPAddress(ip_int))
+                    result.append(ip_attrs)
```

Comment:
The encoding on the wire is designed for the case where many IP addresses share the same set of attributes & handle.  Joining the data to have each unique IP in a list might make the result structure very large.  We should think carefully about downstream applications consuming this data---what is a useful format for them?

Use case you've mentioned: report number of IPs with particular attributes (e.g. "orchestrator": "mesos").  

Any others?